lib/spack/spack/database.py

```diff
@@ -1621,15 +1621,32 @@ def query_local(self, *args, **kwargs):
     query_local.__doc__ += _QUERY_DOCSTRING
 
     def query(self, *args, **kwargs):
-        """Query the Spack database including all upstream databases."""
+        """Query the Spack database including all upstream databases.
+
+        Additional Arguments:
+            install_tree (str): query 'all' (default), 'local', 'upstream', or named upstream
+        """
+        install_tree = kwargs.pop("install_tree", "all")
+        valid_trees = ["all", "upstream", "local", self.root] + [u.root for u in self.upstream_dbs]
+        if install_tree not in valid_trees:
+            msg = "Invalid install_tree argument to Database.query()\n"
+            msg += f"Try one of {', '.join(valid_trees)}"
+            tty.error(msg)
+            return []
```

Comment:
I think this should take *either* a path or a name, as we allow you to name an upstream in config.  The dictionary has keys for each upstream path, so you should have a name on hand.  I know there is no `upstream list` (yet?), but  the user has to write that file and name the trees, and the command help says "upstream name" not "upstream path" right now, and `all` and `local` are names while for other upstreams this only takes paths.  For mirrors we take either -- I think upstreams should be the same.